Mathlib/Analysis/Convex/Deriv.lean

```diff
@@ -366,3 +371,483 @@ theorem strictConcaveOn_univ_of_deriv2_neg {f : ℝ → ℝ} (hf : Continuous f)
     (hf'' : ∀ x, deriv^[2] f x < 0) : StrictConcaveOn ℝ univ f :=
   strictConcaveOn_of_deriv2_neg' convex_univ hf.continuousOn fun x _ => hf'' x
 #align strict_concave_on_univ_of_deriv2_neg strictConcaveOn_univ_of_deriv2_neg
+
+/-!
+## Convexity of `f` implies monotonicity of `f'`
+
+In this section we prove inequalities relating derivatives of convex functions to slopes of secant
+lines, and deduce that if `f` is convex then its derivative is monotone (and similarly for strict
+convexity / strict monotonicity).
+-/
+
+namespace ConvexOn
+
+variable {S : Set ℝ} {f : ℝ → ℝ} {x y f' : ℝ}
+
+section left
+/-!
+### Convex functions, derivative at left endpoint of secant
+-/
+
+/-- If `f : ℝ → ℝ` is convex on `S` and right-differentiable at `x ∈ S`, then the slope of any
+secant line with left endpoint at `x` is bounded below by the right derivative of `f` at `x`. -/
+lemma le_slope_of_hasDerivWithinAt_Ioi (hfc : ConvexOn ℝ S f)
+    (hx : x ∈ S) (hy : y ∈ S) (hxy : x < y) (hf' : HasDerivWithinAt f f' (Ioi x) x) :
+    f' ≤ slope f x y := by
+  apply le_of_tendsto <| (hasDerivWithinAt_iff_tendsto_slope' not_mem_Ioi_self).mp hf'
+  simp_rw [eventually_nhdsWithin_iff, slope_def_field]
+  filter_upwards [eventually_lt_nhds hxy] with t ht (ht' : x < t)
+  refine hfc.secant_mono hx (?_ : t ∈ S) hy ht'.ne' hxy.ne' ht.le
+  exact hfc.1.ordConnected.out hx hy ⟨ht'.le, ht.le⟩
+
+/-- Reformulation of `ConvexOn.le_slope_of_hasDerivWithinAt_Ioi` using `derivWithin`. -/
+lemma right_deriv_le_slope (hfc : ConvexOn ℝ S f) (hx : x ∈ S) (hy : y ∈ S) (hxy : x < y)
+    (hfd : DifferentiableWithinAt ℝ f (Ioi x) x) :
+    derivWithin f (Ioi x) x ≤ slope f x y :=
+  le_slope_of_hasDerivWithinAt_Ioi hfc hx hy hxy hfd.hasDerivWithinAt
+
+/-- If `f : ℝ → ℝ` is convex on `S` and differentiable within `S` at `x`, then the slope of any
+secant line with left endpoint at `x` is bounded below by the derivative of `f` within `S` at `x`.
+
+This is fractionally weaker than `ConvexOn.le_slope_of_hasDerivWithinAt_Ioi` but simpler to apply
+under a `DifferentiableOn S` hypothesis. -/
+lemma le_slope_of_hasDerivWithinAt (hfc : ConvexOn ℝ S f) (hx : x ∈ S) (hy : y ∈ S) (hxy : x < y)
+    (hf' : HasDerivWithinAt f f' S x) :
+    f' ≤ slope f x y := by
+  refine hfc.le_slope_of_hasDerivWithinAt_Ioi hx hy hxy (hf'.mono_of_mem ?_)
+  rw [mem_nhdsWithin_Ioi_iff_exists_Ioc_subset]
+  exact ⟨y, hxy, Ioc_subset_Icc_self.trans (hfc.1.ordConnected.out hx hy)⟩
```

Comment:
I still think this (and the reverse version) could be its own lemma about `ordConnected` sets since you use it at least twice.

Reply:
Happy to do so; but the natural place to put it is `Topology.Order.Basic`, which is already > 2000 lines long, so it will probably generate a linter error. Shall I do a quick PR splitting up that file, and then we can merge this afterwards?

Reply:
I think you can just raise the indicated line limit in `style_exceptions.txt` if that is indeed a problem. If you want to split that file that would be great of course, but I don't think the intention behind these limits is to restrict putting lemmas in the proper places.